sevenzipjbindings/src/main/java/com/mucommander/sevenzipjbindings/multivolume/SevenZipMultiVolumeCallbackHandler.java

```diff
@@ -0,0 +1,92 @@
+package com.mucommander.sevenzipjbindings.multivolume;
+
+import com.mucommander.commons.file.AbstractFile;
+import com.mucommander.commons.file.FileFactory;
+import com.mucommander.sevenzipjbindings.SignatureCheckedRandomAccessFile;
+import net.sf.sevenzipjbinding.*;
+import net.sf.sevenzipjbinding.impl.RandomAccessFileInStream;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.Closeable;
+import java.io.FileNotFoundException;
+import java.io.IOException;
+import java.io.RandomAccessFile;
+import java.util.HashMap;
+import java.util.Map;
+
+public class SevenZipMultiVolumeCallbackHandler implements IArchiveOpenVolumeCallback, ICryptoGetTextPassword, Closeable {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(SevenZipMultiVolumeCallbackHandler.class);
+
+    private Map<String, IInStream> fileCache = new HashMap<>();
+
+    private final AbstractFile firstFile;
+
+    private final byte[] signature;
+
+    private final String password;
+
+    private boolean isFirstFile = true;
+
+    public SevenZipMultiVolumeCallbackHandler(byte[] signature, AbstractFile firstFile, String password) {
+        this.signature = signature;
+        this.firstFile = firstFile;
+        this.password = password;
+    }
+
+    @Override
+    public Object getProperty(PropID propID) throws SevenZipException {
+        switch (propID) {
+            case NAME:
+                return firstFile.getAbsolutePath();
+        }
+        return null;
+    }
+
+    @Override
+    public IInStream getStream(String filename) throws SevenZipException {
+        try {
+            IInStream stream = fileCache.get(filename);
+            if (stream != null) {
+                stream.seek(0, ISeekableStream.SEEK_SET);
+            } else {
+                AbstractFile abstractFile = FileFactory.getFile(filename);
+                if (isFirstFile) {
+                    // Only first file starts with magic number
+                    stream = new SignatureCheckedRandomAccessFile(abstractFile, signature);
+                    isFirstFile = false;
+                } else {
+                    stream = new RandomAccessFileInStream(new RandomAccessFile(filename, "r"));
+                }
+                fileCache.put(filename, stream);
+            }
+            return stream;
+        } catch (FileNotFoundException e) {
+            // There is no way to know ahead of time if we reached the last file,
+            // So it is safe to ignore this Exception
+            return null;
```

Comment:
what happens when we have only some parts available? can it be that the content of an entry is split across several volumes and so we get corrupted data?
if we have to have all the parts then we need to raise an error in this case (maybe even if the data is not corrupted but only some part of the content can be retrieved)

Reply:
If any part is missing, the seven zip library will trigger another exception and the app falls back to the regular error logic. It looks like this:

I renamed the 2nd file to simulate missing volume:
![image](https://github.com/mucommander/mucommander/assets/7936008/23b3113b-03b6-46eb-87fb-c74a129edb80)

And this is what I get when trying to extract:
![image](https://github.com/mucommander/mucommander/assets/7936008/b5a7bf20-504f-4e70-b9c9-bbae084832f6)

Reply:
Ack, that's good